In [54]:
import pandas as pd
ex1 = pd.read_csv('ex1.csv')
ex1 = ex1.rename(index=int, columns={"Unnamed: 0": "Value"})
diff = -0.014048786592986806
per_diff = -1.8152497147529003
diff = round(diff, 3)
per_diff = round(per_diff, 3)

## How to use this document
You can skip the code blocks unless you want to examine a particular calculation. The text slowly walks through a single table (`ex1.csv`) which shows a worked example of what exactly happens to surfaced hits for strikers and non-strikers during a 30% strike of ML-1M.

Some useful context to begin: for ML-1M, the difference in surfaced hits between SVD (fully personalized) and MovieMean (un-personalized) is 0.014, or 1.4% *of all* surfaced hits. This is a -1.815 *percent change* in surfaced hits.

In [55]:
diff, per_diff

(-0.014, -1.815)

First, let's look at the baseline surfaced hits values for our strikers, non-strikers, and the whole system ("all").

In [35]:
ex1[ex1.index == 0]

Value  participants  non-participants     all
0  1) baseline        23.227            54.167  77.393

### Interpreting these baseline values
Without any strike action:
* the participants of the strike (30% of random users) were responsible for 23.23 of surfaced hits
* the non-participants were responsible for 54.17
* together these add up to about 77.4

In [58]:
23.227 + 54.167

77.394

In [48]:
23.227 / 77.393 * 100

30.011758169343484

Indeed, the strikers were contributing about 30% of surfaced hits in the pre-strike condition.

Next, let's look at the next row in this table, which will show how surfaced hits change after the strike.

### Surfaced Hits after strike

In [51]:
ex1[ex1.index.isin([0,1])]

Value  participants  non-participants     all
0           1) baseline        23.227            54.167  77.393
1  2) hits after strike        22.770            53.917  76.687

Looks like surfaced hits decreased a bit for each group after the strike. We may also want to look directly at the *change* in hits.
### Change in Surfaced Hits

In [38]:
ex1[ex1.index.isin([0,1,2])]

Value  participants  non-participants     all
0           1) baseline        23.227            54.167  77.393
1  2) hits after strike        22.770            53.917  76.687
2     3) change in hits        -0.456            -0.250  -0.706

Calculating the change in surfaced hits, it becomes more obvious that the strikers themselves experience more loss in hits than the non-strikers. Recall that strikers are getting un-personalized recommendations. This is a key take-away from this worked example.

We also verify that the effect on the whole system is equal to the sum of these two components.

Next, it will be useful to think about how this change in hits translate to *loss in personalization*. In other words, how far does the strike take performance towards un-personalized?

### Change in Surfaced Hits, normalized w.r.t to un-personalized results

In [39]:
ex1[ex1.index.isin([0,1,2,3])]

Value  participants  \
0                                        1) baseline        23.227   
1                               2) hits after strike        22.770   
2                                  3) change in hits        -0.456   
3  4) change in hits, normalized w.r.t un-persona...        32.471   

   non-participants     all  
0            54.167  77.393  
1            53.917  76.687  
2            -0.250  -0.706  
3            17.771  50.241

If we recall that the difference between SVD and MovieMean (i.e. the personalization gap) is equal to 1.4% surfaced hits, it makes sense that a loss of 0.25 takes non-participants 17.77% of the way towards un-personalized results.

In [70]:
round(-0.250 / -1.4 * 100, 2)

17.86

One puzzling result: why isn't the change in hits for participants not equal to 100% (since they are getting un-personalized recommendations)? This is because the strikers only comprise 30% of the population and we're looking at total surfaced hits still.

In other words, the 30% of strikers can at most take us 30% of the way towards un-personalized results. If we divide the change in hits for strikers, normalized w.r.t un-personalized results, we see indeed the amount is about 30%.

In [67]:
round(-0.456 / -1.4 * 100, 2)

32.57

To get more insight into what's happening from a group's perspective, it might be interesting to consider *percent change* instead of raw change. Specifically, this is percent change for each group, i.e. the change is relative to the group's baseline.

Critically, for an aggregate metric like surfaced hits, each group's baseline is proportional to it's size.

We can look at row 1) to review what the baselines look like.

In [49]:
ex1[ex1.index.isin([0,1,2,3,4])]

Value  participants  \
0                                        1) baseline        23.227   
1                               2) hits after strike        22.770   
2                                  3) change in hits        -0.456   
3  4) change in hits, normalized w.r.t un-persona...        32.471   
4                          5) percent change in hits        -1.964   

   non-participants     all  
0            54.167  77.393  
1            53.917  76.687  
2            -0.250  -0.706  
3            17.771  50.241  
4            -0.461  -0.912

Finally, we can also think of percent change normalized w.r.t un-personalized results.

In other words, how close to un-personalized results did the strike bring *each individual group*. Here, we'd expect the strikers to experience around > 100% change.

In [50]:
ex1[ex1.index.isin([0,1,2,3,4,5])]

Value  participants  \
0                                        1) baseline        23.227   
1                               2) hits after strike        22.770   
2                                  3) change in hits        -0.456   
3  4) change in hits, normalized w.r.t un-persona...        32.471   
4                          5) percent change in hits        -1.964   
5  6) percent change in hits, normalized w.r.t un...       108.199   

   non-participants     all  
0            54.167  77.393  
1            53.917  76.687  
2            -0.250  -0.706  
3            17.771  50.241  
4            -0.461  -0.912  
5            25.399  50.241

Let's dig into why the normalized % change in hits is not the same as normalized change in hits for non-participants, but it is for all.

Non-participants experienced in 0.461% change in hits. The aggregate difference between fully- and un-personalized results is about 1.8.

So a drop of 0.25% of hits corresponds to the system going 17.77% of the way towards un-personalized.

In [68]:
round(0.25 / 1.4 * 100, 2)

17.86

And a percent change of -0.461 corresponds to non-participants experiencing a change that's 25.4% of the way towards un-personalized.

In [69]:
round(0.461 / 1.815 * 100, 2)

25.4

Finally, we clarify the distinction here. When normalizing change in units of surfaced hits, we're always considering the whole system. The "how far towards un-personalized" question is about how close we are to getting the experience of a totally un-personalized system.

Specifically, the full question reads: "How far does a given group take the system towards un-personalized results?" This question is a bit unintuitive and so the results are not too useful.

To get the total normalized change in units of surfaced hits, we can just add the value for each group.

In [64]:
round(32.471 + 17.771, 2)

50.24

When normalizing percent change, we're only considering one group. The "how far towards un-personalized" question is about how close we are to getting the experience of an un-personalized system *for this group*.

Specifically, the full question here should read: "How far does a given group go towards un-personalized results for that group?"
This is probably the more intuitive question we might ask about change relative to un-personalized results. This is the number we report in Section 5.1.

To get the total normalized percent change, we can perform a weighted sum weighted by the size of each group.

In [63]:
round(108.2 * 0.3 + 25.4 * 0.7, 2)

50.24